src/async_transfer.rs

```diff
@@ -0,0 +1,632 @@
+    }
+}
+
+extern "system" fn transfer_finished<T: UsbContext>(transfer_ptr: *mut libusb_transfer) {
+    if transfer_ptr.is_null() {
+        return;
+    }
+
+    // SAFETY: transfer_ptr is not NULL.
+    let transfer: &mut libusb_transfer = unsafe { &mut *transfer_ptr };
+    let user_data = transfer.user_data;
+
+    if user_data.is_null() {
+        return;
+    }
+
+    // SAFETY: user_data is not NULL and the only user always passes a valid `Arc<Mutex<InnerTransfer<T>>>`.
+    let inner = unsafe { Arc::from_raw(user_data as *mut Mutex<InnerTransfer<T>>) };
+    let mut inner = inner.lock().unwrap();
+
+    inner.status = transfer.status;
+    inner.actual_length = transfer.actual_length;
+
+    if let Some(waker) = inner.waker.take() {
+        waker.wake()
+    }
+}
+
+/// Represents a cancellation token for the [`Transfer`]. This allows the user to cancel the USB
+/// transfer while it is still pending.
+#[derive(Clone)]
+pub struct CancellationToken<T: UsbContext> {
+    inner: Arc<Mutex<InnerTransfer<T>>>,
+}
+
+impl<T: UsbContext> CancellationToken<T> {
+    /// Asynchronously cancels the pending USB transfer. This function returns immediately, but
+    /// this does not indicate that the cancellation is complete. Instead this will unblock the
+    /// task awaiting the [`Transfer`] future and cause it to return [`Error::Interrupted`].
+    pub fn cancel(&self) {
+        let mut inner = self.inner.lock().unwrap();
+        let ptr = inner.as_ptr();
+
+        if inner.status == LIBUSB_TRANSFER_ACTIVE {
+            // SAFETY: the transfer is guarded by `Arc<Mutex<T>>` and we only cancel the transfer
+            // if it is still active.
+            unsafe {
+                libusb_cancel_transfer(ptr);
+            }
+        }
+    }
+}
+
+/// Represents a submitted USB transfer that can be polled until completion, in which case it will
+/// return the ownership of the associated buffer. In the transfer got cancelled, this returns
+/// [`Error::Interrupted`].
+pub struct Transfer<T: UsbContext> {
+    inner: Arc<Mutex<InnerTransfer<T>>>,
+    context: T,
+    buffer: Arc<Mutex<Option<Pin<Box<[u8]>>>>>,
+}
+
+impl<T: UsbContext> Drop for Transfer<T> {
+    fn drop(&mut self) {
+        let mut inner = self.inner.lock().unwrap();
+        let ptr = inner.as_ptr();
+
+        inner.waker = None;
+
+        if inner.status == LIBUSB_TRANSFER_ACTIVE {
+            // SAFETY: the transfer is guarded by `Arc<Mutex<T>>` and we only cancel the transfer
+            // if it is still active.
+            unsafe {
+                libusb_cancel_transfer(ptr);
+            }
+        }
+    }
+}
+
+impl<T: UsbContext> Future for Transfer<T> {
+    type Output = Result<(Vec<u8>, usize), (Vec<u8>, Error)>;
+
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
+        // Poll libusb for any complete events.
+        if let Err(e) = self.context.handle_events(Some(Duration::from_micros(0))) {
```

Comment:
This doesn't belong here.

When the transfer is not ready, it is `poll`'s responsibility to save the waker and arrange for it to be woken when the event loop should poll your future again. You save the waker on the transfer and wake it from libusb's completion callback. But where does libusb check for events and call the completion callback? Yep, that's what `libusb_handle_events` and variants are for. So you need to call it somewhere other than `poll` so that libusb can call the callback, callback can wake the waker, and let the executor know that it has to call `poll`.

You have two options:
1. Require a dedicated background thread that loops calling `libusb_handle_events` in blocking mode.
2. If you don't want a background thread and don't need Windows support, you can do the complicated dance to [get a set of file descriptors from libusb](https://libusb.sourceforge.io/api-1.0/group__libusb__poll.html) in a [thread safe manner](https://libusb.sourceforge.io/api-1.0/libusb_mtasync.html#threadwait), and [arrange](https://docs.rs/tokio/latest/tokio/io/unix/struct.AsyncFd.html) for tokio/mio to poll them. When that reports a ready fd, then call `handle_events` in nonblocking mode.